### PR TITLE
define a loop-free untrue batching rule for `rng_bit_generator`

### LINKED ISSUE #16792: unsafe_rbg + vmap --> 10x slow down
### Description

unsafe_rbg is advertised as the solution for performance issues for rng, but it has surprising pessimization in the presence of vmap, more than 10x in this script, but more like 4x in real use.

The key is `vmap(loss)`, where loss calls the RNG.

This is with `LIBTPU_INIT_ARGS=--xla_tpu_spmd_rng_bit_generator_unsafe=true`
 

Times:
* vmap, no dropout: 0.057
* vmap, dropout, threefry: 0.092
* vmap, dropout, unsafe: 0.736  (!!!!!)
* no vmap, no dropout: 0.057
* no vmap, dropout, threefy: 0.088
* no vmap, dropout, unsafe: 0.075

```
# import functools
import time

import jax
import jax.numpy as jnp
import numpy as onp
from jax.sharding import Mesh, NamedSharding, PartitionSpec


batch_size = 256
seq_len = 2048
embed_size = 256
vocab_size = 2000
num_layers = 20

pdrop = 0.1
USE_VMAP = True
USE_UNSAFE_RBG = True

mesh = Mesh(onp.array(jax.devices()), ("dp",))


if USE_UNSAFE_RBG:
    jax.config.update("jax_default_prng_impl", "unsafe_rbg")
else:
    jax.config.update("jax_threefry_partitionable", True)

with mesh:
    key = jax.random.PRNGKey(0)

    def model(tokens, key):
        embed = jnp.take(jnp.ones((vocab_size, embed_size)), tokens, axis=0)
        # dumb fake gpt2 attn
        for i in range(0, num_layers):
            attn = jnp.einsum("...ld,...kd->...lk", embed, embed)

            if pdrop > 0.0:
                key, subkey = jax.random.split(key)
                dout = jax.random.bernoulli(subkey, pdrop, shape=attn.shape)
                attn = jnp.where(dout, jnp.zeros_like(attn), attn)

            attn = jax.nn.softmax(attn, axis=-1)
            embed = jnp.einsum("...ld,...lk->...kd", attn, embed)

        out = jnp.einsum("...ld,...kd->...lk", embed, jnp.ones((vocab_size, embed_size)))

        return out

    def compute_loss(example, key):
        pred_y = model(example, key=key)
        return jnp.mean(pred_y)

    def compute_loss_vmap(examples, key):
        key = jax.random.split(key, batch_size)
        per_ex_loss = jax.vmap(compute_loss)(examples, key)
        return jnp.mean(per_ex_loss)

    if USE_VMAP:
        compute_loss_pjit = jax.jit(compute_loss_vmap)
    else:
        compute_loss_pjit = jax.jit(compute_loss)

    # i still honestly find the way to turn a "replicated" array like batch into a sharded array to be a bit confusing
    batch = jnp.ones((batch_size, seq_len), dtype=jnp.int32)
    batch = jax.make_array_from_callback(
        (batch_size, seq_len), NamedSharding(mesh, PartitionSpec("dp", None)), lambda idx: batch[idx]
    )

    total_loss = 0.0
    total_time = 0.0

    for n in range(20):
        this_key, key = jax.random.split(key)
        time_in = time.time()
        loss = compute_loss_pjit(batch, this_key)

        total_loss += loss.item()
        time_out = time.time()

        if n > 0:
            total_time += time_out - time_in

    print(f"eval loss: {total_loss / n:.3f}")
    print(f"eval time: {total_time / (n-1):.3f}")

```

### What jax/jaxlib version are you using?

0.4.13

### Which accelerator(s) are you using?

TPU

### Additional system info

v3-32

### NVIDIA GPU info

_No response_

### LINKED ISSUE #19085: efficient untrue batching of `random_bit_generator`
The batching rule for the `random_bit_generator` primitive, over a batch of keys, emits a loop (via `lax.map`):
https://github.com/google/jax/blob/42ae8432185bf03f61ddd2e7bc279d3abb5247fd/jax/_src/lax/control_flow/loops.py#L2012-L2024

This is a workaround to the corresponding `RandomBitGenerator` HLO not being batchable. But looping violates the operational expectations of `vmap`, namely that everything is vectorized. And downstream, the surprise performance hit when switching RNG implementations isn't great.

We could consider a few options:
1. Emit an unrolled loop. Drawbacks: grows the program size with the batch size.
2. Generate a batch of random numbers from a _single_ key in the batch, dropping the remaining keys in the batch. Drawbacks: this violates `vmap` semantics considering the random values generated, although the output is "statistically" the same in a sense.

Let's try number 2.

The RBG operation is already non-portable across platforms and XLA flags. In some cases the random generation is affected by sharding. So arguably, callers opting into RBG RNGs already expect unusual semantics. By contrast, it's uncommon that anyone expects the performance hit.

cc @mattjj, @dlwh

### PR DESCRIPTION
**Pull request by froystig**
fixes #19085, #16792

See #19085 for details.

Note that this is a random-bits-altering change, though only for vmapped random generation.

### CODE REVIEW

**Review by mattjj — APPROVED** [review]

**Comment by vikmary** [comment]
This solves my issue with per-example independent sampling (before it was 100X slower compared to dependent sampling), thanks for the fix!
